runners/core-construction-java/src/test/java/org/apache/beam/runners/core/construction/ExecutableStageTranslationTest.java

```diff
@@ -78,6 +81,40 @@ public void processElement(
     String executableStageName =
         ExecutableStageTranslation.generateNameFromStagePayload(basePayload);
 
-    assertThat(executableStageName, is("[3]{ParDo(Anonymous), MyName, count}"));
+    assertThat(executableStageName, is("[3]{ParDo(Anonymous), MyName, Composite}"));
+  }
+
+  @Test
+  public void testOperatorNameGenerationFromNames() {
+    assertGeneratedNames("A", "A", Arrays.asList("A"));
+    assertGeneratedNames("A/a1", "A/a1", Arrays.asList("A/a1"));
+    assertGeneratedNames("A/{a1, a2}", "A/{a1, a2}", Arrays.asList("A/a1", "A/a2"));
+    assertGeneratedNames(
+        "A/{a1, a2}", "A/{a1, a2/{a2.1, a2.2}}", Arrays.asList("A/a1", "A/a2/a2.1", "A/a2/a2.2"));
+    assertGeneratedNames("{A, B}", "{A/{a1, a2}, B}", Arrays.asList("A/a1", "A/a2", "B"));
+    assertGeneratedNames(
+        "{A, B, C}", "{A/{a1, a2}, B, C/c/cc}", Arrays.asList("A/a1", "A/a2", "B", "C/c/cc"));
```

Comment:
The logic is that we only recurse into the composites if we have only one of it?

Reply:
I think that could be unexpected to users because they could assume it either always or never recursively lists the composite transforms. On the other hand, it's good to get a bit more insight for single composite transforms.